File: Day1/Examples/Fortran/gather.f90
!
! Gathering rank ids from all processes at root
!
program gather
    use mpi
    implicit none
    integer (kind=4) :: i,ierror, world_rank, world_size
    integer, allocatable :: buf(:)
  
    call MPI_Init( ierror )
    if (ierror .ne. 0) stop ' Cannot initialize MPI '
    call MPI_Comm_rank( MPI_COMM_WORLD, world_rank, ierror )
    call MPI_Comm_size( MPI_COMM_WORLD, world_size, ierror )

    allocate(buf(0:world_size-1))

    call MPI_Gather( world_rank, 1, MPI_INTEGER, buf, 1, MPI_INTEGER, 0, MPI_COMM_WORLD, ierror )
    if ( world_rank .eq. 0 )  then
            write(6,'(a,i0,a,i0)') 'I am process ',world_rank,' of ',world_size
        do i = 0,world_size-1
            write(6,'(i2)',advance='NO')  buf(i)
        end do
        write(6,*)
    endif

    call MPI_Finalize(ierror);
end program gather

File: Day1/Examples/C/gather.c
/*
 *
 * Gathering rank ids from all processes at root
 *
*/
#include <stdio.h>
#include <mpi.h>

int main( int argc, char **argv ) {
  int myRank, uniSize, ierror;
  int i;

  ierror=MPI_Init(&argc, &argv);
  ierror=MPI_Comm_size(MPI_COMM_WORLD, &uniSize);
  ierror=MPI_Comm_rank(MPI_COMM_WORLD, &myRank);
  
  int *buf= (int*)malloc( sizeof( int ) * uniSize );

  ierror = MPI_Gather( &myRank, 1, MPI_INT, buf, 1, MPI_INT, 0, MPI_COMM_WORLD );
  if(myRank==0){
    printf("Rank 0 printing ...\n");
    for(i=0; i<uniSize; i++){
      printf("From rank %d: %d\n", i, buf[i]);
    }
  }

  ierror = MPI_Finalize();
  return ierror;
}

